### PR TITLE
add RunWithContext + remove signal cancellation

### PR DESCRIPTION
**Pull request by marwan-at-work**
Fixes https://github.com/urfave/cli/issues/945

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -224,7 +232,7 @@ func (a *App) Run(arguments []string) (err error) {
 
 	err = parseIter(set, a, arguments[1:], shellComplete)
 	nerr := normalizeFlags(a.Flags, set)
-	context := NewContext(a, set, nil)
+	context := NewContext(a, set, &Context{Context: ctx})
```
Comment (AudriusButkevicius) [inline app.go]: Will this not blow.up if user did `RunWithContext(nil, args)`?

Comment (AudriusButkevicius) [inline app.go]: Will this still work of user does RunWithContext(nil, args)

File: app.go [hunk 2]
```diff
@@ -207,6 +208,13 @@ func (a *App) useShortOptionHandling() bool {
 // Run is the entry point to the cli app. Parses the arguments slice and routes
 // to the proper flag/args combination
 func (a *App) Run(arguments []string) (err error) {
+	return a.RunWithContext(context.Background(), arguments)
+}
+
+// RunWithContext is like Run except it takes a Context that will be
+// passed to its commands and sub-commands. Through this, you can
+// propagate timeouts and cancellation requests
+func (a *App) RunWithContext(ctx context.Context, arguments []string) (err error) {
```
Comment (AudriusButkevicius) [inline app.go]: Looking at go net lib, this should probably be RunContext

File: app.go (hunk 1 above)
Comment (rliebz) [reply app.go]: Passing a nil Context is always a developer error.

[From the docs](https://golang.org/pkg/context/):

> Do not pass a nil Context, even if a function permits it. Pass context.TODO if you are unsure about which Context to use. 

We shouldn't need to support that

**Comment by marwan-at-work** [comment]
@rliebz @AudriusButkevicius PR updated :)

**Pull request merged**